pkg/controller/mutationtemplate/mutationtemplate_controller.go

```diff
@@ -124,30 +127,104 @@ func (r *ReconcileMutationTemplate) Reconcile(request reconcile.Request) (reconc
 		if errors.IsNotFound(err) {
 			// Object not found, return.  Created objects are automatically garbage collected.
 			// For additional cleanup logic use finalizers.
+			log.Info("didn't fine the instance - finalizer logic here?")
 			return reconcile.Result{}, nil
 		}
 		// Error reading the object - requeue the request.
-		return reconcile.Result{}, err
+		log.Info("some other error occurred in getting the instance, requeuing")
+		return reconcile.Result{Requeue: true}, err
 	}
 	log.Info("Reconciling", "MutationTemplate", instance)
 
 	crd, err := r.opa.CreateMutationCRD(context.Background(), instance)
 	if err != nil {
-		log.Info("CreateMutationCRD failed, apparently.", "error", fmt.Sprintf("%v",err))
+		log.Info("CreateMutationCRD failed, apparently.", "error", fmt.Sprintf("%v", err))
+		return reconcile.Result{}, nil
+	}
+
+	name := crd.GetName()
+	namespace := crd.GetNamespace()
+	if instance.GetDeletionTimestamp().IsZero() {
+		// Check if the constraint already exists
+		found := &apiextensionsv1beta1.CustomResourceDefinition{}
+		err = r.Get(context.TODO(), types.NamespacedName{Name: name, Namespace: namespace}, found)
+		if err != nil && errors.IsNotFound(err) {
+			return r.handleCreate(instance, crd)
+		}
+		log.Info("didn't create or delete")
+		return reconcile.Result{}, err //currently no handleUpdate logic, but it'd go here
+
 	}
 
-	//we are temporarily assuming you're creating a new template and mutation crd
-	return r.handleCreate(instance, crd)
+	log.Info("Deletion timestamp of this", "object", instance.GetDeletionTimestamp())
+
+	return r.handleDelete(instance, crd)
 }
 
 func (r *ReconcileMutationTemplate) handleCreate(instance *templatesv1alpha1.MutationTemplate, crd *apiextensionsv1beta1.CustomResourceDefinition) (reconcile.Result, error) {
+	name := crd.GetName()
+	log := log.WithValues("name", name)
+	log.Info("creating constraint")
+	if !containsString(finalizerName, instance.GetFinalizers()) {
+		instance.SetFinalizers(append(instance.GetFinalizers(), finalizerName))
+		if err := r.Update(context.Background(), instance); err != nil {
+			log.Error(err, "update error", err)
+			return reconcile.Result{Requeue: true}, nil
+		}
+	}
 	log.Info("creating mutation CRD")
 	if err := r.Create(context.TODO(), crd); err != nil {
 		return reconcile.Result{}, err
 	}
 	instance.Status.Created = true
+	log.Info("this is a handleCreate update")
 	if err := r.Update(context.Background(), instance); err != nil {
 		return reconcile.Result{Requeue: true}, nil
 	}
 	return reconcile.Result{}, nil
 }
+
+func (r *ReconcileMutationTemplate) handleDelete(instance *templatesv1alpha1.MutationTemplate, crd *apiextensionsv1beta1.CustomResourceDefinition) (reconcile.Result, error) {
+	name := crd.GetName()
```

Comment:
This is used rarely enough that `crd.GetName()` is probably more clear throughout (ditto elsewhere)